BrainHarmonics/docs/man/man3/Line.3
.TH "Line" 3 "Mon Apr 20 2020" "Version 0.1" "BrainHarmonics" \" -*- nroff -*-
.ad l
.nh
.SH NAME
Line
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <line\&.h>\fP
.PP
Inherits \fBUniverse\fP\&.
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBLine\fP ()"
.br
.ti -1c
.RI "\fBLine\fP (unsigned int object_type)"
.br
.ti -1c
.RI "\fBLine\fP (unsigned int object_type, std::chrono::time_point< \fBClock\fP > event_time)"
.br
.ti -1c
.RI "\fBLine\fP (unsigned int object_type, std::chrono::time_point< \fBClock\fP > event_time, \fBUniverse\fP &universe_connector)"
.br
.ti -1c
.RI "virtual \fB~Line\fP ()"
.br
.ti -1c
.RI "bool \fBResetParameters\fP (std::chrono::time_point< \fBClock\fP > event_time)"
.br
.ti -1c
.RI "int \fBGetLineID\fP (std::chrono::time_point< \fBClock\fP > event_time)"
.br
.ti -1c
.RI "double \fBGetLineXYBubble\fP (std::chrono::time_point< \fBClock\fP > event_time) const"
.br
.ti -1c
.RI "double \fBGetLineX1\fP (std::chrono::time_point< \fBClock\fP > event_time) const"
.br
.ti -1c
.RI "double \fBGetLineY1\fP (std::chrono::time_point< \fBClock\fP > event_time) const"
.br
.ti -1c
.RI "double \fBGetLineX1Y1Bubble\fP (std::chrono::time_point< \fBClock\fP > event_time) const"
.br
.ti -1c
.RI "double \fBGetLineX2\fP (std::chrono::time_point< \fBClock\fP > event_time) const"
.br
.ti -1c
.RI "double \fBGetLineY2\fP (std::chrono::time_point< \fBClock\fP > event_time) const"
.br
.ti -1c
.RI "double \fBGetLineX2Y2Bubble\fP (std::chrono::time_point< \fBClock\fP > event_time) const"
.br
.ti -1c
.RI "double \fBGetLineXMin\fP (std::chrono::time_point< \fBClock\fP > event_time)"
.br
.ti -1c
.RI "double \fBGetLineYMin\fP (std::chrono::time_point< \fBClock\fP > event_time)"
.br
.ti -1c
.RI "double \fBGetLineXMax\fP (std::chrono::time_point< \fBClock\fP > event_time)"
.br
.ti -1c
.RI "double \fBGetLineYMax\fP (std::chrono::time_point< \fBClock\fP > event_time)"
.br
.ti -1c
.RI "double \fBGetLineMinOrdinal\fP (std::chrono::time_point< \fBClock\fP > event_time)"
.br
.ti -1c
.RI "double \fBGetLineMaxOrdinal\fP (std::chrono::time_point< \fBClock\fP > event_time)"
.br
.ti -1c
.RI "double \fBGetSlope\fP (std::chrono::time_point< \fBClock\fP > event_time)"
.br
.ti -1c
.RI "double \fBGetYint\fP (std::chrono::time_point< \fBClock\fP > event_time)"
.br
.ti -1c
.RI "double \fBGetLineTTL\fP (std::chrono::time_point< \fBClock\fP > event_time) const"
.br
.ti -1c
.RI "int \fBGetLineSeq\fP (std::chrono::time_point< \fBClock\fP > event_time)"
.br
.ti -1c
.RI "int \fBGetLineCounter\fP (std::chrono::time_point< \fBClock\fP > event_time) const"
.br
.ti -1c
.RI "bool \fBGetInLine\fP (std::chrono::time_point< \fBClock\fP > event_time, double X1, double Y1, double X2, double Y2, double Threshold)"
.br
.ti -1c
.RI "\fBvoid\fP \fBsetLineID\fP (std::chrono::time_point< \fBClock\fP > event_time, int \fBval\fP)"
.br
.ti -1c
.RI "\fBvoid\fP \fBsetLineX1\fP (std::chrono::time_point< \fBClock\fP > event_time, double \fBval\fP)"
.br
.ti -1c
.RI "\fBvoid\fP \fBsetLineY1\fP (std::chrono::time_point< \fBClock\fP > event_time, double \fBval\fP)"
.br
.ti -1c
.RI "\fBvoid\fP \fBsetLineX2\fP (std::chrono::time_point< \fBClock\fP > event_time, double \fBval\fP)"
.br
.ti -1c
.RI "\fBvoid\fP \fBsetLineY2\fP (std::chrono::time_point< \fBClock\fP > event_time, double \fBval\fP)"
.br
.ti -1c
.RI "\fBvoid\fP \fBsetSlope\fP (std::chrono::time_point< \fBClock\fP > event_time, double \fBval\fP)"
.br
.ti -1c
.RI "\fBvoid\fP \fBsetYint\fP (std::chrono::time_point< \fBClock\fP > event_time, double \fBval\fP)"
.br
.ti -1c
.RI "\fBvoid\fP \fBsetLineTTL\fP (std::chrono::time_point< \fBClock\fP > event_time, double \fBval\fP)"
.br
.ti -1c
.RI "\fBvoid\fP \fBsetLineSeq\fP (std::chrono::time_point< \fBClock\fP > event_time, int \fBval\fP)"
.br
.ti -1c
.RI "\fBvoid\fP \fBsetLineCounter\fP (std::chrono::time_point< \fBClock\fP > event_time, int \fBval\fP)"
.br
.ti -1c
.RI "int \fBUpdate\fP (std::chrono::time_point< \fBClock\fP > event_time)"
.br
.in -1c
.SS "Additional Inherited Members"
.SH "Detailed Description"
.PP 
< For Sine, Cosine, Fabs & Sqrt functions 
.PP
Definition at line 25 of file line\&.h\&.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "Line::Line ()\fC [inline]\fP"

.PP
Definition at line 28 of file line\&.h\&.
.SS "Line::Line (unsigned int object_type)\fC [inline]\fP"

.PP
Definition at line 30 of file line\&.h\&.
.SS "Line::Line (unsigned int object_type, std::chrono::time_point< \fBClock\fP > event_time)\fC [inline]\fP"

.PP
Definition at line 32 of file line\&.h\&.
.SS "Line::Line (unsigned int object_type, std::chrono::time_point< \fBClock\fP > event_time, \fBUniverse\fP & universe_connector)\fC [inline]\fP"

.PP
Definition at line 34 of file line\&.h\&.
.SS "virtual Line::~Line ()\fC [inline]\fP, \fC [virtual]\fP"
Default destructor 
.PP
Definition at line 47 of file line\&.h\&.
.SH "Member Function Documentation"
.PP 
.SS "bool Line::GetInLine (std::chrono::time_point< \fBClock\fP > event_time, double X1, double Y1, double X2, double Y2, double Threshold)\fC [inline]\fP"

.PP
Definition at line 73 of file line\&.h\&.
.SS "int Line::GetLineCounter (std::chrono::time_point< \fBClock\fP > event_time) const\fC [inline]\fP"

.PP
Definition at line 72 of file line\&.h\&.
.SS "int Line::GetLineID (std::chrono::time_point< \fBClock\fP > event_time)\fC [inline]\fP"

.PP
Definition at line 54 of file line\&.h\&.
.SS "double Line::GetLineMaxOrdinal (std::chrono::time_point< \fBClock\fP > event_time)\fC [inline]\fP"

.PP
Definition at line 67 of file line\&.h\&.
.SS "double Line::GetLineMinOrdinal (std::chrono::time_point< \fBClock\fP > event_time)\fC [inline]\fP"

.PP
Definition at line 66 of file line\&.h\&.
.SS "int Line::GetLineSeq (std::chrono::time_point< \fBClock\fP > event_time)\fC [inline]\fP"

.PP
Definition at line 71 of file line\&.h\&.
.SS "double Line::GetLineTTL (std::chrono::time_point< \fBClock\fP > event_time) const\fC [inline]\fP"

.PP
Definition at line 70 of file line\&.h\&.
.SS "double Line::GetLineX1 (std::chrono::time_point< \fBClock\fP > event_time) const\fC [inline]\fP"

.PP
Definition at line 56 of file line\&.h\&.
.SS "double Line::GetLineX1Y1Bubble (std::chrono::time_point< \fBClock\fP > event_time) const\fC [inline]\fP"

.PP
Definition at line 58 of file line\&.h\&.
.SS "double Line::GetLineX2 (std::chrono::time_point< \fBClock\fP > event_time) const\fC [inline]\fP"

.PP
Definition at line 59 of file line\&.h\&.
.SS "double Line::GetLineX2Y2Bubble (std::chrono::time_point< \fBClock\fP > event_time) const\fC [inline]\fP"

.PP
Definition at line 61 of file line\&.h\&.
.SS "double Line::GetLineXMax (std::chrono::time_point< \fBClock\fP > event_time)\fC [inline]\fP"

.PP
Definition at line 64 of file line\&.h\&.
.SS "double Line::GetLineXMin (std::chrono::time_point< \fBClock\fP > event_time)\fC [inline]\fP"

.PP
Definition at line 62 of file line\&.h\&.
.SS "double Line::GetLineXYBubble (std::chrono::time_point< \fBClock\fP > event_time) const\fC [inline]\fP"

.PP
Definition at line 55 of file line\&.h\&.
.SS "double Line::GetLineY1 (std::chrono::time_point< \fBClock\fP > event_time) const\fC [inline]\fP"

.PP
Definition at line 57 of file line\&.h\&.
.SS "double Line::GetLineY2 (std::chrono::time_point< \fBClock\fP > event_time) const\fC [inline]\fP"

.PP
Definition at line 60 of file line\&.h\&.
.SS "double Line::GetLineYMax (std::chrono::time_point< \fBClock\fP > event_time)\fC [inline]\fP"

.PP
Definition at line 65 of file line\&.h\&.
.SS "double Line::GetLineYMin (std::chrono::time_point< \fBClock\fP > event_time)\fC [inline]\fP"

.PP
Definition at line 63 of file line\&.h\&.
.SS "double Line::GetSlope (std::chrono::time_point< \fBClock\fP > event_time)\fC [inline]\fP"

.PP
Definition at line 68 of file line\&.h\&.
.SS "double Line::GetYint (std::chrono::time_point< \fBClock\fP > event_time)\fC [inline]\fP"

.PP
Definition at line 69 of file line\&.h\&.
.SS "bool Line::ResetParameters (std::chrono::time_point< \fBClock\fP > event_time)"
Access lineCounter 
.PP
\fBReturns:\fP
.RS 4
The current value of lineCounter 
.RE
.PP

.PP
Definition at line 19 of file line\&.cc\&.
.SS "\fBvoid\fP Line::setLineCounter (std::chrono::time_point< \fBClock\fP > event_time, int val)\fC [inline]\fP"

.PP
Definition at line 94 of file line\&.h\&.
.SS "\fBvoid\fP Line::setLineID (std::chrono::time_point< \fBClock\fP > event_time, int val)\fC [inline]\fP"
set lineCounter 
.PP
\fBParameters:\fP
.RS 4
\fIval\fP New value to set 
.RE
.PP

.PP
Definition at line 85 of file line\&.h\&.
.SS "\fBvoid\fP Line::setLineSeq (std::chrono::time_point< \fBClock\fP > event_time, int val)\fC [inline]\fP"

.PP
Definition at line 93 of file line\&.h\&.
.SS "\fBvoid\fP Line::setLineTTL (std::chrono::time_point< \fBClock\fP > event_time, double val)\fC [inline]\fP"

.PP
Definition at line 92 of file line\&.h\&.
.SS "\fBvoid\fP Line::setLineX1 (std::chrono::time_point< \fBClock\fP > event_time, double val)\fC [inline]\fP"

.PP
Definition at line 86 of file line\&.h\&.
.SS "\fBvoid\fP Line::setLineX2 (std::chrono::time_point< \fBClock\fP > event_time, double val)\fC [inline]\fP"

.PP
Definition at line 88 of file line\&.h\&.
.SS "\fBvoid\fP Line::setLineY1 (std::chrono::time_point< \fBClock\fP > event_time, double val)\fC [inline]\fP"

.PP
Definition at line 87 of file line\&.h\&.
.SS "\fBvoid\fP Line::setLineY2 (std::chrono::time_point< \fBClock\fP > event_time, double val)\fC [inline]\fP"

.PP
Definition at line 89 of file line\&.h\&.
.SS "\fBvoid\fP Line::setSlope (std::chrono::time_point< \fBClock\fP > event_time, double val)\fC [inline]\fP"

.PP
Definition at line 90 of file line\&.h\&.
.SS "\fBvoid\fP Line::setYint (std::chrono::time_point< \fBClock\fP > event_time, double val)\fC [inline]\fP"

.PP
Definition at line 91 of file line\&.h\&.
.SS "int Line::Update (std::chrono::time_point< \fBClock\fP > event_time)"

.PP
Definition at line 59 of file line\&.cc\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for BrainHarmonics from the source code\&.


BrainHarmonics/line.cc
/*------------------------------------------------------------------------------*/
/* @file      line.cc                                                           */
/* @details   Assign structure for scalar values                                */
/* @author    Paul Isaac's                                                      */
/* @date      16.02.2016                                                        */
/* @Copyright © 2016 Paul Isaac's. All rights reserved.                         */
/*------------------------------------------------------------------------------*/

/* Code snippets used:                                                          */
/* Syntax comparison - http://stackoverflow.com & http://cplusplus.com          */
/* The class defines a data point and the operations that can be carried out on */
/* it.                                                                          */
/* Using the hierarchical linking the aim is to develop the application to      */
/* relate to real-world physics. This will then ease mapping between simulation,*/
/* emulation and real-world universes.                                          */

#include "line.h"

bool Line::ResetParameters(std::chrono::time_point<Clock> event_time)
{
        // If object instantiated without a time specified then add one.
    if(time_object_created == std::chrono::time_point<Clock>(std::chrono::nanoseconds::zero()))
        {
        time_object_created = event_time;
        srand((std::chrono::duration_cast<std::chrono::seconds>(TheTimeNow().time_since_epoch()).count()));
        }
    
        // If object instantiated without a type specified the set one.
    if(line_type < 1)
        {
        line_type = 1;
        }
    
    object_energy = 10 * ((rand() % 30 + 85) / 100);
    object_energy_threshold = 2000 * ((rand() % 30 + 85) / 100);
    object_size = 1 * ((rand() % 30 + 85) / 100);
    
    switch(line_type)
    {
        case 0:
        {
        object_size = 3 * ((rand() % 30 + 85) / 100);
        break;
        }
        case 1:
        {
        object_size = 2 * ((rand() % 30 + 85) / 100);
        break;
        }
        case 2:
        {
        object_size = 1 * ((rand() % 30 + 85) / 100);
        break;
        }
    }
    return true;
}

int Line::Update(std::chrono::time_point<Clock> event_time)
{
        // If this is the first time that Update is called after object instantiation use the setup parameters.
        // object_initialised should then report true and then not be re-run.
    if(!object_initialised)
        {
        object_initialised = ResetParameters(event_time);
        }
    
    if(event_time != previous_event_time)
        {
    AdjustCounters(event_time);
    
    duration_since_last_event = std::chrono::duration_cast<std::chrono::nanoseconds>(event_time - previous_event_time).count();
    if (duration_since_last_event < 0)
        {
        duration_since_last_event = 0;
        }
    
    if (object_energy < 0)
        {
        object_energy = 0;
        }
    
    if (duration_since_last_event > 0 && object_energy > 0)
        {
            // Extra programming: Change 0 to correct point address
            // Use GetPoint to enable movement across InterneuronSpace to Neuroreceptor
        // time_dimension_pointer->AddTemporalAdjustment(event_time, &GetPoint(event_time, 0)->object_energy, object_energy * 0.1, 10000, 1);
        // time_dimension_pointer->AddTemporalAdjustment(event_time, &object_energy, 0 - (object_energy * 0.1), 10000, 1);
        }
        previous_event_time = event_time;
        }
    return 0;
}
